### PR TITLE
Respect original text decorator order

### LINKED ISSUE #4795: export default is moved after decorator
**Prettier 1.13.7**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAdKcAeAHCAnGAAgBM4AzAQwFcAbIgAUigzBgApgBfASnTBooBnQYQCimCgFtsNOISzwoxEQCU4FVgDoAwhGnQERYOkKE8CUnkIBeQm27WAfISi0a6TiAA0ICNhgAltCCyKAUeHgQAO4ACuEIISgUAG4QAcTeIABGeBoA1nAwAMrYGgFQAObIMHhUcD7lgnAEMbkVkhTIlDRNPgBWgpgAQrlgBcVScAAy5XBdFD31IKV4TXjI2RRZAJ400JnYeOUwAOrpMAAWyAAcAAw+hxBNJ7nYG4dwa8lzPuYAjlQAuZWhR2p0kN1eiAmpIAtVaktBOUKrIAIpUCDweaLHwwLZnYiXZAAJlxuQCNGRukkHQ2UAMmSoTQAKltEpC4JxOEA)
```sh
--parser babylon
```

**Input:**
```jsx

export default @connect({})
class Example extends React.Component {
  render = ()=> null
}
```

**Output:**
```jsx
@connect({})
export default class Example extends React.Component {
  render = () => null;
}

```

**Expected behavior:**
It should keep export default where I put it.

### LINKED ISSUE #5045: Support new decorators proposal
https://github.com/babel/babel-eslint/releases/tag/v9.0.0

**Input:**
```jsx
import React, { PureComponent } from 'react';

import { connect } from 'dva';

export default
@connect(({ list }) => ({
  list,
}))
class Center extends PureComponent {
  render() {
    return <div>123</div>
  }
}

```

**Output:**
```jsx
import React, { PureComponent } from 'react';

import { connect } from 'dva';

@connect(({ list }) => ({
  list,
}))
export default class Center extends PureComponent {
  render() {
    return <div>123</div>;
  }
}

```

**Expected behavior:**
Using the export keyword between a decorator and a class is not allowed. Please use `export @dec class` instead.

prettier/prettier-vscode#572

### PR DESCRIPTION
**Pull request by duailibe**
Closes #5045
Closes #4795
Supersedes #2630

Hopefully this will enable to continue using Prettier without too much problems until the spec is standardized

### CODE REVIEW

**Comment by vjeux** [comment]
I don't have the time to do a proper review but I like what it does.

File: src/language-js/parser-babylon.js [hunk 1]
```diff
@@ -4,71 +4,99 @@ const createError = require("../common/parser-create-error");
...
-      "optionalCatchBinding",
-      "optionalChaining",
-      "classPrivateProperties",
-      ["pipelineOperator", { proposal: "minimal" }],
-      "nullishCoalescingOperator",
-      "bigInt",
-      "throwExpressions"
-    ]
-  };
+  const combinations = [
```
Comment (ikatyang) [inline src/language-js/parser-babylon.js]: We probably need a `guess` function (similar to `isProbablyJsx` in `parser-typescript.js`) to try not to do unnecessary parsing.

File: tests/typescript_decorators/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -49,8 +49,7 @@ export class TabCompletionController {}
   selector: "angular-component"
 })
 class AngularComponent {
-  @Input()
-  myInput: string;
+  @Input() myInput: string;
```
Comment (ikatyang) [inline tests/typescript_decorators/__snapshots__/jsfmt.spec.js.snap]: This looks like the changes from #5188, rebasing issue?

File: src/language-js/parser-babylon.js (hunk 1 above)
Comment (duailibe) [reply src/language-js/parser-babylon.js]: I thought about it, trying to find a `@deco` before/after an export

File: tests/typescript_decorators/__snapshots__/jsfmt.spec.js.snap (hunk 2 above)
Comment (duailibe) [reply tests/typescript_decorators/__snapshots__/jsfmt.spec.js.snap]: Oh probably

**Comment by duailibe** [comment]
I'm making the call and merging this. We can fix any bugs that show up after the release. Will try a solution for the "guess decorator before or after export" in a follow-up PR if nobody beats me to it

**Comment by existentialism** [comment]
@duailibe sorry for the delay, been an insane week. Agree with rolling with this and fixing any bugs.

**Pull request merged**